jopa-impl/src/main/java/cz/cvut/kbss/jopa/model/metamodel/EntityLifecycleCallbackResolver.java

```diff
@@ -51,8 +51,8 @@ class EntityLifecycleCallbackResolver {
     EntityLifecycleListenerManager resolve() {
         resolveLifecycleCallbacks();
         resolveEntityListeners();
-        if (managedType.getSupertype() != null) {
-            manager.setParent(managedType.getSupertype().getLifecycleListenerManager());
+        if (managedType.getSupertypes() != null) {
+            managedType.getSupertypes().forEach(supertype -> manager.addParent(((AbstractIdentifiableType<?>) supertype).getLifecycleListenerManager()));
```

Comment:
I would change the signature of `AbstractidentifiableType.getSupertypes` to return a set of `AbstractIdentifiableType`s (which corresponds to the field anyway). This should prevent the need for the cast here.

Reply:
I actually tried to do this during implementation, however it is not as straightforward. 
I cannot change the `AbstractIdentifiableType.getSupertypes` to return `Set<AbstractIdentifiableType<? super X>>`, because it is not type compatible with `IdentifiableType.getSupertypes`. This is because Java generics are not covariant - see [this](https://docs.oracle.com/javase/tutorial/extra/generics/subtype.html) .

And I cannot change the signature of `IdentifiableType.getSupertypes` because `IdentifiableType` is in `jopa-api` and AIT is in `jopa-impl` module.

Reply:
Ok, could you go ahead a try changing the signature of the method in `IdentifiableType` to `Set<? extends IdentifiableType<? super X>> getSupertypes()`. It breaks a bunch of tests when I try it but it seems mostly to be a problem of mocking the IdentifiableType interface...

Reply:
The signature was changed, tests modified, pointless retyping removed. All test are passing, thanks for the proposal !